Add support for more colors in ncurses output. WIP for #91

hashhar opened the pull request:
WIP for #91 

**NOTE: This breaks the `b` and `c` keys for cycling through the colors
in ncurses mode.**

Added a new function in cava.c `int validate_color(char
*colourToValidate, int outputMode)` which returns 1 if the colour is
valid, 0 otherwise.

Validation is done to ensure that HTML colours are of the form `#xxxxxx` or a
named ncurses colour is present.

HTML colours are allowed only when output mode is ncurses (ie `om = 1 ||
om = 2`). HTML and named colours can be mixed.

If the output mode is other than ncurses, the old method of setting
`bgcol` and `col` will be used.

Added a function in `terminal_ncurses.c` called `struct cols
parse_color(char *col)` which returns a struct containing:
1. `col`: The ncurses colour number parsed.
2. `R`, `G` and `B`: Integers that represent the R, G and B values
   parsed.

The `init_terminal_ncurses()` method now accepts two character pointers
instead of integers. The pointers contain color names (either HTML or
ncurses names) which are then parsed by the parse function defined
above.

Currently there is no check to find if the terminal supports color
redefinition or not. According to the parsed value we decide whether to
perform redefinition or use the predefined ncurses colours.

Ncurses colours 1 and 2 are redefined for foreground and background
respectively. This means that upon exiting, all the occurrences of 1 and
2 would have been redefined. To fix the colours back to the original the
user has to open another terminal.

hashhar commented:
TODO:

- [ ] Add a check for deciding if color redefinition is possible or not.
- [ ] Fix the broken `b` and `c` key functionality.
- [ ] Write tests and modularize the code to make it easier to read and reduce chances of regressions.

hashhar commented:
I'm currently upgrading my system so I may not be able to do any work for the day. 😞

karlstav commented:
@hashhar I finally got around to check this out. Looks really really good! Also it looks like it's actually 24-bit? Since ncurses is not using the colors already defined by the terminal, but redefining new colors instead.

Do you think I should merge it yet? Or maybe create a separate branch for it?

hashhar commented:
I made some changes. I have yet to push them. I will try to push them tomorrow.

I added code to check if the terminal supports color redefinition or not and simplified the code a little bit.

But I was not able to fix the broken `b` and `c` keys feature. I think you should create a separate branch for the time being while I try to fix those issues.

karlstav commented:
created new branch `more-color`, I don't know if I can "move" this PR or if we have to close this and you create a new PR to the new branch?

hashhar commented:
Hey, I think is good enough for now. Maybe you should merge this and we can see if it has any issues for people who want to try it.

The only things that will not work are:

1. `b` and `c` keys to cycle between colors. I am trying to fix it but time is an issue.
2. No errors are currently displayed if the terminal doesn't support color redefiniton, I simply fallback to the default colours in that case.

karlstav commented:
some thoughts (if interested):

 1. maybe just disable color cycling when a custom color is defined? (I don't think it will be missed)
 2. an error and exit would be nice, so people don't just get mad over it not working

I might look into it over the weekend

hashhar commented:
Thanks. I'll add the error and exit and I think I'll disable the color cycling for now.
